### DISCUSSION TITLE
How to run pytest with authentication

### DISCUSSION THREAD

**Discussion by junoriosity** [fastapi-users/fastapi-users#830] [Q&A]
I would like to run pytest on several APIs I have written. Essentially they look as follow:

```
@router.post("/", response_model=SomeSchemaDB, status_code=201)
async def create_project(payload: SomeSchema, user: User = Depends(current_user)):
     SOME CODE
```

And in it, I basically just interact with a Postgresql DB to do someCRUD operations.

My question is now ihow to do some simple testing with pytest that does not run into problems because of the `current_user`authentication.

For instance, I could just have some `conftest.py`like

```
import pytest
from starlette.testclient import TestClient

from app.main import app


@pytest.fixture(scope="module")
def test_app():
    client = TestClient(app)
    yield client  # testing happens here
```

and something like `test_schema.py` which looks like

```
import json

import pytest

from app.api import crud


def test_create_note(test_app, monkeypatch):
    test_request_payload = {"title": "something", "description": "something else"}
    test_response_payload = {"id": 1, "title": "something", "description": "something else"}

    async def mock_post(payload):
        return 1

    monkeypatch.setattr(crud, "post", mock_post)

    response = test_app.post("/notes/", data=json.dumps(test_request_payload),)

    assert response.status_code == 201
    assert response.json() == test_response_payload
```
(For, reference, this testing part is taken from https://testdriven.io/blog/fastapi-crud/).

So, in particular, I would like to make use of monkeypatch to mock some functions.

I would know how to do this without any authentication. However, since we have the `current_user`part I do not know how to handle this here.

Hence, it would be great if someone could help me here.

**Comment by frankie567** [comment]
The easiest way is to use [`dependency_overrides` mechanism](https://fastapi.tiangolo.com/advanced/testing-dependencies) to override the return value of your `current_user` dependency.

```py
test_user = UserDB(email="test@email.com", hashed_password="aaa")

@pytest.fixture(scope="module")
def test_app():
    app.dependency_overrides[current_user] = lambda: test_user
    client = TestClient(app)
    yield client  # testing happens her
```

**Reply by junoriosity** [reply]
@frankie567 Many thanks, I manage now to run some tests. However, I frequently want to test how users can exchange items.

Consider, for instance, the following situation: 

- User A, wants to add user B to some project
- User B must be in the database, whereas user A must be the current user
- Say, now user B wants to remove user A from the project
- then user A must be in the database and user B must be the current user

Do you have some idea how to implement it along these lines?

**Reply by frankie567** [reply]
It'll greatly depend on your setup and project.

To do this kind of things, you'll generally need to insert some data in your database before your unit tests (in a pytest fixture for example).

This way, you can fill your database with the users and related objects you need and then test the behaviour of your API.
